### ISSUE TITLE
Pascal VOC2007 performance is not as good

### ISSUE THREAD

**Issue by TheShadow29** [yhenon/pytorch-retinanet#31]
Hi. First of all thanks for your clean and clear repository. 

I am trying to evaluate the models on VOC2007, however, I am unable to reach the required performance. SSD with resnet50 backbone on pascal voc achieves 79.7% map (reported here:https://github.com/ShuangXieIrene/ssds.pytorch#performance). However, in my experiments it is always saturating at 55mAP. I also tried with resnet18 and resnet34. The first saturates around 50, the second goes upto 55 but saturates thereafter.

The only possible difference is the input image size. I believe you are resizing images with image size of 608. I highly doubt that is indeed causing problem.

Would like to know your thoughts on this.

**Comment by yhenon** [comment]
I'm not sure exactly what the underlying reason is. I know others have the same issue when trying to train retinanet on pascal voc. I suspect retinanet works less well with the smaller dataset.

>Also could you share your parameters to train on coco2017? In particular, how many epochs did you run it for.

I trained around 40 epochs, with the default parameters.

**Comment by TheShadow29** [comment]
How long did it take for training? For me each epoch is taking around 6-7 hours, 40 epochs would mean around 6 days. I am using GTX 1080ti. Is there any suggestion you have to make the training faster?

**Comment by yhenon** [comment]
Unfortunately not much can be done about training speed - more GPUs and/or better GPUs help. Would be interested to hear about performance with voc07+12, although I've heard from others that retinanet performance on pascal voc is lacking, and may require tuning of the alpha and gamma parameters.

**Comment by TheShadow29** [comment]
Unfortunate news. The mAP stagnates at 57-58%. The only difference I had from the current implementation is having patience for Reduce LR to be 2 instead of 3. Doubt that is causing the problem. 

A possible reason would be less data augmentation. Here https://github.com/amdegroot/ssd.pytorch#map the author states that he gets upto 58% without data augmentation, and 77% with data augmentation. I believe the author is performing quite a few data augmentations while this repo does random flip.

**Comment by qq184861643** [comment]
@TheShadow29 @yhenon 
Hi, sorry for @ you guys but I have a question in code and I think it maybe related to this issue.
At model.py class ResNet, in its forward() function, line 279 it write as
`nms_scores, nms_class = classification[0, anchors_nms_idx, :].max(dim=1)`

why in evaluation mode this only return the first batch of results?
shouldn't it be`nms_scores, nms_class = classification[:, anchors_nms_idx, :].max(dim=1)` ?

**Comment by TheShadow29** [comment]
@qq184861643 Sorry for the delayed reply. I think in evaluation it is one image at a time (batch size is 1), there is no dataloader, rather iteration is over the dataset. Hence `classification[0, anchors_nms_idx, :]` is equivalent to `classification[:, anchors_nms_idx, :]`

**Comment by qq184861643** [comment]
@TheShadow29 Yeah I found it, still thx for your replay!

**Comment by WangLuluTju** [comment]
> 
> 
> Hi. First of all thanks for your clean and clear repository.
> 
> I am trying to evaluate the models on VOC2007, however, I am unable to reach the required performance. SSD with resnet50 backbone on pascal voc achieves 79.7% map (reported here:https://github.com/ShuangXieIrene/ssds.pytorch#performance). However, in my experiments it is always saturating at 55mAP. I also tried with resnet18 and resnet34. The first saturates around 50, the second goes upto 55 but saturates thereafter.
> 
> The only possible difference is the input image size. I believe you are resizing images with image size of 608. I highly doubt that is indeed causing problem.
> 
> Would like to know your thoughts on this.
> 
> Also could you share your parameters to train on coco2017? In particular, how many epochs did you run it for.

**Comment by WangLuluTju** [comment]
@TheShadow29 hi, sorry to trouble you. I want to train the model on VOC dataset, but I have trouble with the code. Could you please send me the updated code with interface processing VOC dataset? My email is huachuang@tju.edu.cn

**Comment by pengfeidip** [comment]
> Unfortunate news. The mAP stagnates at 57-58%. The only difference I had from the current implementation is having patience for Reduce LR to be 2 instead of 3. Doubt that is causing the problem.
> 
> A possible reason would be less data augmentation. Here https://github.com/amdegroot/ssd.pytorch#map the author states that he gets upto 58% without data augmentation, and 77% with data augmentation. I believe the author is performing quite a few data augmentations while this repo does random flip.

I don't think so. Because in original paper, in addition to flipping, no other data augmentations.

**Comment by SHU-FLYMAN** [comment]
I think maybe it is wrong in this part:
![image](https://user-images.githubusercontent.com/38397845/61992467-7cad7600-b091-11e9-84ed-6186d711d433.png)

The image format of `skimage.io.imread` is H, W, C, but this padding part is seem to W, H, C.

**Comment by Wuziyi616** [comment]
@SHU-FLYMAN Hi, do you get a better performance after changing the code you mentioned? Thanks!

**Comment by tiger-k** [comment]
my result base on yolov5-6.0-yolov5s,without use Pre-trained models, just voc2007train+voc2007val  map on voc2007val is 0.54.
batch_size 64
